torch/csrc/jit/serialization/callstack_debug_info_serialization.cpp

```diff
@@ -0,0 +1,189 @@
+#include <torch/csrc/jit/api/compilation_unit.h>
+#include <torch/csrc/jit/serialization/callstack_debug_info_serialization.h>
+#include <torch/csrc/jit/serialization/pickle.h>
+
+namespace torch {
+namespace jit {
+
+c10::IValue InlinedCallStackSerializer::serialize(
+    const InlinedCallStackPtr& cs_ptr,
+    const SourceRangeTagMap& source_range_tags) {
+  if (!cs_ptr) {
+    return c10::IValue();
+  }
+  auto cs_it = serialized_inlined_callstack_.find(cs_ptr);
+  if (cs_it != serialized_inlined_callstack_.end()) {
+    return cs_it->second;
+  }
+  // Inlined callstack pointer is serialized as tuple of 3 elements
+  // {IValue(module_instance_info), source_range_tag, IValue(InlinedCallStack)}
+  std::vector<c10::IValue> elements;
+  elements.reserve(3);
+  elements.emplace_back(
+      serialize_module_instance_info(cs_ptr->module_instance()));
+  int64_t source_range_tag{-1};
+  const SourceRange& sr = cs_ptr->source_range().findSourceRangeThatGenerated()
+      ? cs_ptr->source_range().findSourceRangeThatGenerated().value()
+      : cs_ptr->source_range();
+  auto sr_it = source_range_tags.find(sr);
+  if (sr_it != source_range_tags.end()) {
+    source_range_tag = sr_it->second;
+  }
+  elements.emplace_back(source_range_tag);
+  if (cs_ptr->callee()) {
+    elements.emplace_back(
+        serialize(cs_ptr->callee().value(), source_range_tags));
```

Comment:
So this will recursively serialize the whole chain of callstack in denormalized form?
Have we considered some form of normalization, i.e. saving some unique ID of 'cs_ptr->callee()' instead of saving the entire chain recursively? Does module_instance_info + source_range_tag uniquely identify an InlinedCallStackPtr node? Can we use them (or their hash) as the unique ID?

Or it doesn't matter in practice because this won't bloat the model size?

Reply:
Looks like 'Pickler::pushIValue" does memorize the PtrType IValue - so it would be automatically 'normalized' by the pickler?

Reply:
Yes, this will be automatically normalized off of the cached IValue that we return. It is similar to how source range is serialized, https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/serialization/source_range_serialization.cpp